contracts/StorageFees.cdc

```diff
@@ -0,0 +1,180 @@
+import FlowToken from "./FlowToken.cdc"
+
+pub contract StorageFees {
+
+    // Defines the minimum unit (chunk) size of storage in bytes. Storage can only be bought (or refunded) in a multiple of the minimum storage unit.
+    pub var minimumStorageUnit: UInt64
+
+    // Defines the minimum amount of `storage_capacity` an address can have and also the amount every new account has. `minimumAccountStorage` is a multiple of `minimumStorageUnit`.
+    pub var minimumAccountStorage: UInt64
+
+    // Defines the cost of 1 byte of storage in FLOW tokens.
+    pub var flowPerByte: UFix64
+
+    // Defines the cost of purchasing the initial minimum storage in FLOW tokens.
+    pub var flowPerAccountCreation: UFix64
+
+    // Defines the cost of 1 byte of storage in FLOW tokens.
+    pub var refundingEnabled: Bool
+
+    access(contract) var idCounter: UInt64
+
+    pub resource Administrator {
+        pub fun setRefundingEnabled(enabled: Bool) {
+            StorageFees.refundingEnabled = enabled
+        }
+
+        pub fun setMinimumAccountStorage(minimumAccountStorage: UInt64) {
+            pre{
+                minimumAccountStorage % StorageFees.minimumStorageUnit == UInt64(0): "Minimum account storage must be a multiple of the minimum storage unit"
+            }
+            StorageFees.minimumAccountStorage = minimumAccountStorage
+        }
+
+        pub fun setFlowPerByte(flowPerByte: UFix64) {
+            StorageFees.flowPerByte = flowPerByte
+        }
+
+        pub fun setFlowPerAccountCreation(flowPerAccountCreation: UFix64) {
+            StorageFees.flowPerAccountCreation = flowPerAccountCreation
+        }
+    }
+
+    pub resource interface StorageCapacityCapability{
+        access(contract) var id: UInt64
+        pub var storageCapacity: UInt64
+        access(contract) let address: Address
+        access(contract) fun addStorageCapacity(address: Address, amount: UInt64, paymentVault: @FlowToken.Vault)
+    }
+
+    pub struct StorageCapacityPurchase{
+        pub let storageCapacity: UInt64
+        pub let flowCost: UFix64
+
+        init(storageCapacity: UInt64, flowCost: UFix64) {
+            self.storageCapacity = storageCapacity
+            self.flowCost = flowCost
+        }
+    }
+
+    pub resource StorageCapacity: StorageCapacityCapability {
+        access(contract) var id: UInt64
+        pub var storageCapacity: UInt64
+        access(contract) let address: Address
+        access(self) let vault: @FlowToken.Vault
+        access(self) let purchases: [StorageCapacityPurchase]
+
+        access(contract) fun addStorageCapacity(address: Address, amount: UInt64, paymentVault: @FlowToken.Vault){
+            pre{
+                self.address == address: "Unexpected address on storage capacity"
+            }
+            self.purchases.append(StorageCapacityPurchase(storageCapacity: amount, flowCost: paymentVault.balance))
```

Comment:
I'm really confused as to why we need to keep a record of the individual purchases. This seems to create a lot of bookkeeping overhead, and I'm not sure why that's valuable.

Reply:
The idea was that even if the price of storage changes over time refunding purchased storage would give you back the same amount of flow as was originally used to purchase it.

I purchase 100 kb at 1 mFlow/kb then 50 kb at 2 mFlow/kb. The price now is 3 mFlow/kb. If I refund 100kb I would get back 100 mFlow (for the  50 kb at 2 mFlow/kb) + 50 (for the 100 kb at 1 mFlow/kb).

Is this not necessary? Can we just return at the current rate until there are funds  in the vault? This could lead to tokens being locked: 

I purchase 100 kb at 1 mFlow/kb; this is my only purchase.The price now is 0.1 mFlow/kb. I refund all my storage (down to the minimum) and I get 10 mFlow. There is still 90 mFlow in the vault I cant get to.